### PR TITLE
fix testnet support

### PR DESCRIPTION
**Pull request by skylarbarrera**
I added a toggle in the network switcher since if we enable them by default things are ugly for a normal user (see video) happy to change the UX around this as I think the network switcher has no point for a normal user.

update:
- added https://faucet.paradigm.xyz for clout but also cause its the best

POW(updated): 
https://cloud.skylarbarrera.com/Screen-Recording-2021-11-27-00-43-57.mp4

QA: 
-  make sure the options are consistent on iOS and android.
-  make sure the setting persists between sessions.
-  [example.walletconnect.org](example.walletconnect.org) should show networks switching properly
- tx's should work as expected, send & walletconnect

### CODE REVIEW

**Comment by skylarbarrera** [comment]
imo we should add a network disclaimer as otherwise there is no visual cue: only issue is we don't have the eth badges necessary to pipe it in properly. cc @christianbaroni 

https://cloud.skylarbarrera.com/Screen-Shot-2021-11-24-01-14-57.88.png

File: src/handlers/localstorage/globalSettings.js [hunk 1]
```diff
@@ -63,3 +64,20 @@ export const saveTransactionSignatures = transactionSignatures =>
  */
 export const getTransactionSignatures = () =>
   getGlobal(TRANSACTION_SIGNATURES, {});
+
+/**
+ * @desc get dev enabled preference
+ * @param  {Boolean}  [value]
+ */
+export const getTestnetsEnabled = async preference => {
```
Comment (jinchung) [inline src/handlers/localstorage/globalSettings.js:72]: `preference` should not be passed in here

File: src/handlers/localstorage/globalSettings.js [hunk 2]
```diff
@@ -63,3 +64,20 @@ export const saveTransactionSignatures = transactionSignatures =>
...
+  const test = getGlobal(TESTNETS_ENABLED, preference);
```
Comment (jinchung) [inline src/handlers/localstorage/globalSettings.js:73]: the second param is the default value that we should return for when the key TESTNETS_ENABLED cannot be found (for instance, on the very first time a user reads this key before setting any future value)

File: src/handlers/localstorage/globalSettings.js (hunk 1 above)
Comment (skylarbarrera) [reply src/handlers/localstorage/globalSettings.js:72]: oops yeah i did a refactor/name change and missed this

File: src/handlers/localstorage/globalSettings.js (hunk 2 above)
Comment (jinchung) [inline src/handlers/localstorage/globalSettings.js:73]: can follow the pattern of the other functions and just do:
```
export const getTestnetsEnabled = async preference =>
  getGlobal(TESTNETS_ENABLED, preference);
```

Remember that `getGlobal` is an async function so line 73 might not be doing what you expect, though in this case we're just passing the promise back up

File: src/handlers/localstorage/globalSettings.js [hunk 3]
```diff
@@ -63,3 +64,20 @@ export const saveTransactionSignatures = transactionSignatures =>
...
+  return test;
+};
+
+/**
+ * @desc save dev  enabled preference
+ * @param  {Boolean}  [value]
+ */
+export const saveTestnetsEnabled = async preference => {
```
Comment (jinchung) [inline src/handlers/localstorage/globalSettings.js:81]: you can remove `async` here

File: src/redux/settings.js [hunk 4]
```diff
@@ -53,6 +61,29 @@ export const settingsLoadNetwork = () => async dispatch => {
...
+    const testnetsEnabled = await getTestnetsEnabled();
+    dispatch({
+      payload: testnetsEnabled,
+      type: SETTINGS_UPDATE_TESTNET_PREF_SUCCESS,
+    });
+  } catch (error) {
+    logger.log('Error loading testnets preference', error);
+  }
+};
+export const settingsChangeTestnetsEnabled = testnetsEnabled => async dispatch => {
```
Comment (jinchung) [inline src/redux/settings.js:75]: space pls 🙏

File: src/redux/settings.js [hunk 5]
```diff
@@ -53,6 +61,29 @@ export const settingsLoadNetwork = () => async dispatch => {
   }
 };
 
+export const settingsLoadTestnetsEnabled = () => async dispatch => {
```
Comment (jinchung) [inline src/redux/settings.js:64]: u don't need this function yes?

File: src/redux/settings.js [hunk 6]
```diff
@@ -53,6 +61,29 @@ export const settingsLoadNetwork = () => async dispatch => {
...
+      payload: testnetsEnabled,
+      type: SETTINGS_UPDATE_TESTNET_PREF_SUCCESS,
+    });
+  } catch (error) {
+    logger.log('Error loading testnets preference', error);
+  }
+};
+export const settingsChangeTestnetsEnabled = testnetsEnabled => async dispatch => {
+  saveTestnetsEnabled(testnetsEnabled);
+  try {
```
Comment (jinchung) [inline src/redux/settings.js:77]: not sure why you need a try/catch around this

Comment (jinchung) [reply src/redux/settings.js:77]: also we generally have a pattern of dispatching to redux before saving to localstorage

File: src/screens/TransactionConfirmationScreen.js [hunk 7]
```diff
@@ -237,6 +238,10 @@ export default function TransactionConfirmationScreen() {
     return isL2Network(network);
   }, [network]);
 
+  const isTestnet = useMemo(() => {
```
Comment (jinchung) [inline src/screens/TransactionConfirmationScreen.js:241]: this is an example of something where bc it is so lightweight, the memo becomes more overhead than help

can just be:
```
const isTestnet = isTestnetNetwork(network);
```
(and yes we're doing this inconsistently across app :( )

Comment (skylarbarrera) [reply src/screens/TransactionConfirmationScreen.js:241]: gonna change isL2 as well

File: src/screens/TransactionConfirmationScreen.js [hunk 8]
```diff
@@ -245,11 +246,12 @@ export default function TransactionConfirmationScreen() {
 
   useEffect(() => {
     const initProvider = async () => {
-      const p = isL2 ? await getProviderForNetwork(network) : web3Provider;
+      const p =
```
Comment (jinchung) [inline src/screens/TransactionConfirmationScreen.js:249]: oh now that we have `mainnet` set up in this provider cache, we can actually get rid of this check and just straight up use `getProviderForNetwork` here

Comment (jinchung) [reply src/screens/TransactionConfirmationScreen.js:249]: and can get rid of using `web3Provider` directly in this file

File: src/components/settings-menu/SettingsSection.js [hunk 9]
```diff
@@ -136,7 +136,11 @@ export default function SettingsSection({
 }) {
   const isReviewAvailable = false;
   const { wallets, isReadOnlyWallet } = useWallets();
-  const { /*language,*/ nativeCurrency, network } = useAccountSettings();
+  const {
+    /*language,*/ nativeCurrency,
```
Comment (brunobar79) [inline src/components/settings-menu/SettingsSection.js:140]: nasty

File: src/screens/TransactionConfirmationScreen.js (hunk 8 above)
Comment (brunobar79) [reply src/screens/TransactionConfirmationScreen.js:249]: Seems like you still need to do this (web3provider still here)

**Review by brunobar79 — APPROVED** [review]

File: src/redux/fallbackExplorer.js [hunk 10]
```diff
@@ -475,10 +475,10 @@ export const fallbackExplorerInit = () => async (dispatch, getState) => {
       },
       type: FALLBACK_EXPLORER_SET_ASSETS,
     });
+    dispatch(explorerInitL2());
```
Comment (skylarbarrera) [inline src/redux/fallbackExplorer.js:478]: polygon assets we're sometimes showing up so I moved this inside the mainnet check

File: src/components/pools/PoolsListWrapper.js [hunk 11]
```diff
@@ -19,9 +20,9 @@ export default function PoolsListWrapper({
     toggleOpenInvestmentCards,
   } = useOpenInvestmentCards();
 
-  if (isCoinListEdited) {
-    return null;
-  }
+  const { network } = useAccountSettings();
+
+  if (isCoinListEdited || isTestnetNetwork(network)) return null;
```
Comment (skylarbarrera) [inline src/components/pools/PoolsListWrapper.js:25]: same thing here, so we don't render

File: src/components/AddFundsInterstitial.js [hunk 12]
```diff
@@ -115,9 +115,8 @@ const AmountButtonWrapper = styled(Row).attrs({
   ${android ? 'width: 100' : ''};
 `;
 
-const onAddFromFaucet = network => {
-  const faucetUrl = get(networkInfo[network], 'faucet_url');
-  Linking.openURL(faucetUrl);
+const onAddFromFaucet = () => {
+  Linking.openURL('https://faucet.paradigm.xyz');
```
Comment (jinchung) [inline src/components/AddFundsInterstitial.js:119]: looks like line 275 is still passing along the `network`
```<InterstitialButton onPress={() => onAddFromFaucet(network)}```

also, you are defining a function inside of that onPress, should be able to just pass `onAddFromFaucet` now 

also you can remove the `{}` like:
```
const onAddFromFaucet = () => Linking.openURL('https://faucet.paradigm.xyz');
```

Comment (skylarbarrera) [reply src/components/AddFundsInterstitial.js:119]: lol oop

Comment (skylarbarrera) [reply src/components/AddFundsInterstitial.js]: fixed

File: src/components/L2Disclaimer.js [hunk 13]
```diff
@@ -46,11 +51,15 @@ const L2Disclaimer = ({
...
-              size="small"
-            />
+            {isL2 ? (
+              <ChainBadge
+                assetType={assetType}
+                position="relative"
+                size="small"
+              />
+            ) : (
+              <CoinIcon address="eth" size={20} symbol="ETH" />
```
Comment (jinchung) [inline src/components/L2Disclaimer.js:61]: `ETH_ADDRESS` from `references`

File: src/components/AddFundsInterstitial.js [hunk 14]
```diff
@@ -273,7 +271,7 @@ const AddFundsInterstitial = ({ network }) => {
             faucet
           </Title>
           <Row marginTop={30}>
-            <InterstitialButton onPress={() => onAddFromFaucet(network)}>
+            <InterstitialButton onPress={onAddFromFaucet(accountAddress)}>
```
Comment (jinchung) [inline src/components/AddFundsInterstitial.js:274]: ah now that it has an argument again we will need to do the `() => onAddFromFaucuet(accountAddress)` - can pull it out as a const function so it isn't defined inline

**Comment by jinchung** [comment]
Status update: still have a few files to review

File: src/redux/settings.js [hunk 15]
```diff
@@ -116,6 +133,12 @@ export const INITIAL_STATE = {
 
 export default (state = INITIAL_STATE, action) => {
```
Comment (jinchung) [inline src/redux/settings.js:134]: I don't see an initial state for `testnetsEnabled` in lines 126-132

File: src/hooks/useAccountSettings.js [hunk 16]
```diff
@@ -24,13 +25,22 @@ export default function useAccountSettings() {
   const { language } = useSelector(createLanguageSelector);
   const dispatch = useDispatch();
   const settingsData = useSelector(
-    ({ settings: { accountAddress, chainId, nativeCurrency, network } }) => ({
+    ({
+      settings: {
+        accountAddress,
+        chainId,
+        testnetsEnabled,
```
Comment (jinchung) [inline src/hooks/useAccountSettings.js:32]: abc

File: src/components/settings-menu/SettingsSection.js [hunk 17]
```diff
@@ -244,20 +249,22 @@ export default function SettingsSection({
...
-            </ListItemArrowGroup>
-          </ListItem>
+          {testnetsEnabled && (
+            <ListItem
+              icon={
+                <SettingIcon
+                  source={isDarkMode ? NetworkIconDark : NetworkIcon}
+                />
+              }
+              label="Network"
```
Comment (jinchung) [inline src/components/settings-menu/SettingsSection.js:259]: lang.t

File: src/helpers/walletConnectNetworks.js [hunk 18]
```diff
@@ -10,9 +11,12 @@ const androidReverseNetworkWithName = name =>
 
 export const NETWORK_MENU_ACTION_KEY_FILTER = 'switch-to-network-';
 
-export const networksMenuItems = () =>
-  Object.values(networkInfo)
-    .filter(({ disabled, testnet }) => !disabled && !testnet)
+export const networksMenuItems = () => {
+  const testnetsEnabled = store.getState().settings.testnetsEnabled;
```
Comment (jinchung) [inline src/helpers/walletConnectNetworks.js:15]: `const { testnetsEnabled } = store.getState().settings;`

File: src/screens/TransactionConfirmationScreen.js (hunk 7 above)
Comment (jinchung) [reply src/screens/TransactionConfirmationScreen.js:225]: this comment still stands

File: src/screens/TransactionConfirmationScreen.js [hunk 19]
```diff
@@ -234,24 +238,22 @@ export default function TransactionConfirmationScreen() {
...
     const initProvider = async () => {
-      const p = isL2
-        ? await getProviderForNetwork(currentNetwork)
-        : web3Provider;
+      const p = await getProviderForNetwork(currentNetwork);
       setProvider(p);
     };
     currentNetwork && initProvider();
-  }, [isL2, currentNetwork]);
+  }, [isL2, isTestnet, currentNetwork]);
```
Comment (jinchung) [inline src/screens/TransactionConfirmationScreen.js:245]: this useEffect no longer needs `isL2` in the dependency list

File: src/redux/settings.ts [hunk 20]
```diff
@@ -106,6 +121,33 @@ export const settingsLoadNetwork = () => async (
   }
 };
 
+export const settingsLoadTestnetsEnabled = () => async (
```
Comment (jinchung) [inline src/redux/settings.ts:124]: this function isn't called anywhere - you can just do the dispatch as part of `settingsLoadState` - you may want to create a new action for "load success" that updates both the native currency and the testnets pref, and just delete this function.

File: src/handlers/localstorage/globalSettings.ts [hunk 21]
```diff
@@ -64,3 +65,16 @@ export const saveTransactionSignatures = (transactionSignatures: any) =>
  */
 export const getTransactionSignatures = () =>
   getGlobal(TRANSACTION_SIGNATURES, {});
+
+/**
+ * @desc get testnets enabled preference
+ */
+export const getTestnetsEnabled = () => getGlobal(TESTNETS_ENABLED, false);
```
Comment (jinchung) [inline src/handlers/localstorage/globalSettings.ts:72]: if this is defaulted to false, what happens to the users who had switched to a testnet before upgrading

File: src/handlers/localstorage/globalSettings.ts [hunk 22]
```diff
@@ -64,3 +65,16 @@ export const saveTransactionSignatures = (transactionSignatures: any) =>
...
+
+/**
+ * @desc save testnets enabled preference
+ * @param  {Boolean}  [value]
+ */
+export const saveTestnetsEnabled = (preference: boolean) => {
```
Comment (jinchung) [inline src/handlers/localstorage/globalSettings.ts:78]: also leaving a note here for if this should be in mmkv

File: src/helpers/walletConnectNetworks.js [hunk 23]
```diff
@@ -55,7 +60,7 @@ export const changeConnectionMenuItems = () => {
           iconType: 'SYSTEM',
           iconValue: 'network',
         },
-        menuItems: networksAvailable,
+        menuItems: networksMenuItems(),
```
Comment (jinchung) [inline src/helpers/walletConnectNetworks.js:63]: leaving a review bookmark for myself here

File: src/components/settings-menu/UserDevSection.js [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+import { InteractionManager } from 'react-native';
+import { ScrollView, Switch } from 'react-native-gesture-handler';
+import { useDispatch } from 'react-redux';
+import networkTypes from '../../helpers/networkTypes';
+import { settingsUpdateNetwork } from '../../redux/settings';
+import { Column } from '../layout';
+import { ListFooter, ListItem } from '../list';
+import NetworkSection from './NetworkSection';
+
+import {
```
Comment (jinchung) [inline src/components/settings-menu/UserDevSection.js:11]: can u cleanup the spacing here - remove spacing above and insert a space below this import line

File: src/components/settings-menu/UserDevSection.js [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+    testnetsEnabled,
+    settingsChangeTestnetsEnabled,
+  } = useAccountSettings();
+
+  const resetAccountState = useResetAccountState();
+  const loadAccountData = useLoadAccountData();
+  const initializeAccountData = useInitializeAccountData();
+
+  const revertToMainnet = useCallback(async () => {
+    await resetAccountState();
```
Comment (jinchung) [inline src/components/settings-menu/UserDevSection.js:30]: hey other than updating the network, is there any reason we need to reset the account / reload account data / reinit account data?

File: src/handlers/localstorage/globalSettings.ts (hunk 22 above)
Comment (skylarbarrera) [reply src/handlers/localstorage/globalSettings.ts:78]: feel like we should move all settings related stuff to mmkv at the same time if we're going to that

File: src/components/settings-menu/UserDevSection.js (hunk 25 above)
Comment (skylarbarrera) [reply src/components/settings-menu/UserDevSection.js:30]: i'm following what we're doing in `NetworkSection.js`

**Comment by skylarbarrera** [comment]
fixes RNBW-2459

File: src/handlers/localstorage/globalSettings.ts (hunk 21 above)
Comment (skylarbarrera) [reply src/handlers/localstorage/globalSettings.ts:72]: they will still be in testnets but can still go to settings -> dev settings to fix. imo this is fine, testnets are so broken rn so the user % is prob really small, can be handled by support if they can't find

**Comment by scotato** [comment]
lgtm 👀

**Review by ibrahimtaveras00 — APPROVED** [review]
Fixes look good, QA Passed 👍🏽

**Pull request merged**